### DISCUSSION TITLE
Include Path

### DISCUSSION THREAD

**Discussion by 7rwang** [Homebrew/discussions#5334] [Getting started]
### Output of `brew config`

```shell
HOMEBREW_VERSION: 4.2.20
ORIGIN: https://github.com/Homebrew/brew
HEAD: c2ed3327c605c3e738359c9807b8f4cd6fec09eb
Last commit: 3 days ago
Core tap JSON: 30 Apr 13:20 UTC
Core cask tap JSON: 30 Apr 13:20 UTC
HOMEBREW_PREFIX: /opt/homebrew
HOMEBREW_CASK_OPTS: []
HOMEBREW_MAKE_JOBS: 10
Homebrew Ruby: 3.1.4 => /opt/homebrew/Library/Homebrew/vendor/portable-ruby/3.1.4/bin/ruby
CPU: 10-core 64-bit arm_firestorm_icestorm
Clang: 15.0.0 build 1500
Git: 2.39.3 => /Applications/Xcode.app/Contents/Developer/usr/bin/git
Curl: 8.4.0 => /usr/bin/curl
macOS: 14.2-arm64
CLT: 15.1.0.0.1.1700200546
Xcode: 15.2
Rosetta 2: false
```

### Output of `brew doctor`

```shell
Your system is ready to brew.
```

### Description of issue

Hi everyone,

I'm not particularly familiar with C++ and related stuffs like compiler etc. Recently I'm stuck with an include issue. I just wanna know why I should export the path 'opt/homebrew/include' rather than the library located path. I have checked the located path is 'opt/brew/Cellar'.  As I understand it I think the correct path should be the latter.

Thanks you all in advance!

**Comment by SMillerDev** [comment]
Because `/opt/homebrew/Cellar/curl/8.7.1/include/curl/` will be `/opt/homebrew/Cellar/curl/8.7.2/include/curl/` on the next release. So you'll need to change your paths all the time. Meanwhile `/opt/homebrew/opt/curl/include/curl/` or `/opt/homebrew/include` will always exist and point to the latest version of the formula

**Reply by 7rwang** [reply]
Thx! So `opt/homebrew/opt/curl/include/curl/` and `/opt/homebrew/include` these two pathes are equivalent? BTW why the last is `include` rather than` Cellar`? If I use /opt/homebrew/Cellar it will not point to some libraries?

**Reply by SMillerDev** [reply]
> Thx! So opt/homebrew/opt/curl/include/curl/ and /opt/homebrew/include these two pathes are equivalent? 

They are equivalent for some software. 
- `/opt/homebrew/include` <- this is like `/usr/include` on Linux, it will include everything that made it's include items globally available
- `/opt/homebrew/opt/curl/include/curl/` <- this is the specific include directory for cURL. If you need to include PHP headers, that won't work.
> BTW why the last is include rather than Cellar? If I use /opt/homebrew/Cellar it will not point to some libraries?

The first part of my answer explained why you don't want to use Cellar based paths.

**Reply by 7rwang** [reply]
`/opt/homebrew/include` will make the include items globally available but only **available** for the libraries which are installed by `homebrew` right? 
And if there is an another package mangement tool whose the name maybe `Sometool` and if I wanna export an enviroment variable it will be `opt/Sometool/include`?

**Reply by SMillerDev** [reply]
> `/opt/homebrew/include` will make the include items globally available but only available for the libraries which are installed by homebrew right?

Yes, homebrew can't manage the include paths it does not know about.
> And if there is an another package mangement tool whose the name maybe Sometool and if I wanna export an enviroment variable it will be opt/Sometool/include?

No, it could be anything. There is no guarantee other tools would follow the same naming Homebrew does.

**Reply by 7rwang** [reply]
I see. One more question. If I use `Sometool` like I mentioned above how can I know which path should I export or how can I tell the compiler where can it find the library?

**Reply by SMillerDev** [reply]
You'd have to read their documentation

**Reply by 7rwang** [reply]
Got it! Thank you very much dude. U saved my life.
